### PR TITLE
fix-selection-area-throw-on-scroll

### PR DESCRIPTION
**Pull request by EArminjon**
Many issues are trying to be fixed here :
- Selection context menu and selection hands inside SelectionArea disappear for ever when we scroll the view.
- SelectionArea is broken inside scrollable view and throw an exception because of null value on scroll.
- SelectionArea control's hand in a scrollable are broken when we drag scroll outside the listview current loaded scope.

https://github.com/flutter/flutter/issues/139797
https://github.com/flutter/flutter/issues/124078

<details><summary>Code example</summary>
<p>

```dart
// import 'flavors/france/frx_config.dart';
// import 'main.dart';
//
// void main() => runWithConfig(FrxUatConfig());

import 'package:flutter/material.dart';

void main() => runApp(
      const App(),
    );

class App extends StatelessWidget {
  const App({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  const Home({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: const Text("SelectionArea issue"),
      ),
      body: SafeArea(
        child: SelectionArea(
          child: Column(
            children: <Widget>[
              const TextField(),
              Expanded(
                child: ListView.separated(
                  shrinkWrap: true,
                  padding:
                      const EdgeInsets.symmetric(horizontal: 15, vertical: 10),
                  itemCount: 50,
                  separatorBuilder: (BuildContext context, int index) =>
                      const Divider(height: 1),
                  itemBuilder: (BuildContext context, int index) {
                    return IntrinsicHeight(
                      child: Row(
                        crossAxisAlignment: CrossAxisAlignment.stretch,
                        children: <Widget>[
                          Expanded(
                            child: ColoredBox(
                              color: Colors.grey,
                              child: Padding(
                                padding: const EdgeInsets.all(16),
                                child: Text("key-$index"),
                              ),
                            ),
                          ),
                          Expanded(
                            child: Padding(
                              padding: const EdgeInsets.all(16),
                              child: Text(
                                "value-$index",
                              ),
                            ),
                          ),
                        ],
                      ),
                    );
                  },
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

```
</p>
</details> 

BEFORE : 

Cannot select and scroll
untitled

AFTER : 

Can select and scroll
not-perfect

TODO :

- Once controller hands visible again we need to display back the selection menu ? (I notice that TextFields seems to not do it...)
- Trying to select a large content inside a shrinkWrap:false listview is broken. The selection is inconsistant. Maybe we need to reset the selection, I've trouble to achieve that.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by EArminjon** [comment]
This PR is 'ready' because it fix the exception and so in release user will no see a GREY screen. 
BUT, selectionArea is buguy and this PR is not enough :/

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 1]
```diff
@@ -1208,21 +1207,21 @@ class SelectableRegionState extends State<SelectableRegion> with TextSelectionDe
 
   /// The line height at the start of the current selection.
   double get startGlyphHeight {
-    return _selectionDelegate.value.startSelectionPoint!.lineHeight;
+    return _selectionDelegate.value.startSelectionPoint?.lineHeight ?? 0;
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/selectable_region.dart:1210]: Nice catch, it looks like when either selection edge is not in the viewport then start/endSelectionPoint may be `null` which is causing the crash when scrolling the selection out of view..

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 2]
```diff
@@ -429,8 +429,7 @@ class SelectableRegionState extends State<SelectableRegion> with TextSelectionDe
     if (_hasSelectionOverlayGeometry) {
       _updateSelectionOverlay();
     } else {
-      _selectionOverlay?.dispose();
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/selectable_region.dart:432]: Why is this change needed?

**Review by Renzo-Olivares — COMMENTED** [review]
Hello @EArminjon, thank you for your contribution! These fixes seem like a great improvement to `SelectionArea`. Before this can be merged it will also need a test and existing tests should be passing.

I've tried this PR out and can confirm it fixes:

> SelectionArea is broken inside scrollable view and throw an exception because of null value on scroll.

This one still seems to be broken:

> Selection context menu and selection hands inside SelectionArea disappear for ever when we scroll the view.

And I'm not sure I completely understand this issue:

> SelectionArea control's hand in a scrollable are broken when we drag scroll outside the listview current loaded scope. 

Can you clarify or provide a video?

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 2 above)
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/selectable_region.dart:432]: I suspect it might be causing the failing tests.

**Comment by goderbauer** [comment]
(triage) @EArminjon Just checking in to see if you still have plans to come back to this PR to address the feedback given above?

**Comment by EArminjon** [comment]
Hello,

I will try, very time boxed. Feel free to improve this PR guys, it's not easy to fix

**Comment by goderbauer** [comment]
I am going to close this for now since it is getting stale. It is referenced from the issue https://github.com/flutter/flutter/issues/124078. Feel free to reopen if you find the time to work on this again. Thanks for the work that already went into this!
